bpf/lib/nodeport.h

```diff
@@ -832,6 +833,81 @@ int tail_nodeport_nat_egress_ipv6(struct __ctx_buff *ctx)
 	return send_drop_notify_error_ext(ctx, 0, ret, ext_err, CTX_ACT_DROP, METRIC_EGRESS);
 }
 
+declare_tailcall_if(__not(is_defined(IS_BPF_LXC)), CILIUM_CALL_IPV6_NODEPORT_DSR_INGRESS)
+int tail_nodeport_dsr_ingress_ipv6(struct __ctx_buff *ctx)
+{
+	struct ct_state ct_state_new = {};
+	struct ipv6_ct_tuple tuple = {};
+	struct ct_state ct_state = {};
+	void *data, *data_end;
+	struct ipv6hdr *ip6;
+	__u32 monitor = 0;
+	union v6addr addr;
+	int ret, l4_off;
+	__be16 port;
+
+	if (!revalidate_data(ctx, &data, &data_end, &ip6)) {
+		ret = DROP_INVALID;
+		goto drop_err;
+	}
+
+	ret = lb6_extract_tuple(ctx, ip6, &l4_off, &tuple);
+	if (IS_ERR(ret))
+		goto drop_err;
+
+	addr.p1 = ctx_load_meta(ctx, CB_ADDR_V6_1);
+	addr.p2 = ctx_load_meta(ctx, CB_ADDR_V6_2);
+	addr.p3 = ctx_load_meta(ctx, CB_ADDR_V6_3);
+	addr.p4 = ctx_load_meta(ctx, CB_ADDR_V6_4);
+	port = (__be16)ctx_load_meta(ctx, CB_PORT);
+
+	ctx_store_meta(ctx, CB_PORT, 0);
+	ctx_store_meta(ctx, CB_ADDR_V6_1, 0);
+	ctx_store_meta(ctx, CB_ADDR_V6_2, 0);
+	ctx_store_meta(ctx, CB_ADDR_V6_3, 0);
+	ctx_store_meta(ctx, CB_ADDR_V6_4, 0);
+
+	ret = ct_lb_lookup6(get_ct_map6(&tuple), &tuple, ctx, l4_off,
+			    CT_EGRESS, &ct_state, &monitor);
+	switch (ret) {
+	case CT_NEW:
+	case CT_REOPENED:
+create_ct:
+		ct_state_new.src_sec_id = WORLD_ID;
+		ct_state_new.dsr = 1;
+		ct_state_new.ifindex = (__u16)NATIVE_DEV_IFINDEX;
+		ret = ct_create6(get_ct_map6(&tuple), NULL, &tuple, ctx,
+				 CT_EGRESS, &ct_state_new, false, false, false);
+		if (!IS_ERR(ret))
+			ret = snat_v6_create_dsr(&tuple, &addr, port);
+
+		if (IS_ERR(ret))
+			goto drop_err;
+		break;
+	case CT_ESTABLISHED:
+		if (tuple.nexthdr == IPPROTO_TCP || !ct_state.dsr)
+			goto create_ct;
+		break;
+	case CT_REPLY:
+		ipv6_ct_tuple_reverse(&tuple);
+		goto create_ct;
+	default:
+		ret = DROP_UNKNOWN_CT;
+		goto drop_err;
+	}
+
+	ret = neigh_record_ip6(ctx);
+	if (ret < 0)
+		goto drop_err;
+
+	ctx_skip_nodeport_set(ctx);
+	ep_tail_call(ctx, CILIUM_CALL_IPV6_FROM_NETDEV);
+	ret = DROP_MISSED_TAIL_CALL;
```

Comment:
`nodeport_lb6` used to perform a tail call to `CILIUM_CALL_IPV6_NODEPORT_NAT_INGRESS` after calling `nodeport_uses_dsr6`. Why is this changed to go to `CILIUM_CALL_IPV6_FROM_NETDEV` instead?

Reply:
The code in NAT_INGRESS is only needed if we're dealing with some sort of reply traffic - revSNAT, potentially revDNAT, EgressGW reply .... But if the packet has embedded DSR info or matches a DSR CT entry, we already know it's not reply traffic. So we could circle through NAT_INGRESS as well, but at the moment it's just not needed.

Reply:
Got it. Thanks for the explanation!

Might be worst a note in the commit description, though maybe it's evident to LB folks.

Reply:
added
>If a packet belongs to a DSR connection (ie gets handled by
tail_nodeport_dsr_ingress\_\*()), we know it's not reply traffic. Thus
we don't need to jump through CILIUM_CALL\_\*\_NODEPORT_NAT_INGRESS
afterwards, and can circle straight back to CILIUM_CALL\_\*\_FROM_NETDEV.